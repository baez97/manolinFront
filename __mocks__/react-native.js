function AsyncStorageClass() {
    this.setItem = jest.fn((key, value) => {
        return new Promise((resolve, reject) => {
            resolve();
        })
    });

    this.getItem = jest.fn((key, value) => {
        return new Promise((resolve, reject) => {
            resolve("THIS_IS_THE_SAVED_TOKEN");
        })
    });

    this.removeItem = jest.fn((key) => {
        return  new Promise((resolve, reject) => {
            resolve();
        })
    });

    this.removeItem.mockReturnValueOnce(new Promise((res,rej) => {
        rej({message: "THIS IS AN ERROR"});
    }))
}

module.exports.AsyncStorage = new AsyncStorageClass();