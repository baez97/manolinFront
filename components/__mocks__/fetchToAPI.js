const fetchToAPI = jest.fn();
export default fetchToAPI;