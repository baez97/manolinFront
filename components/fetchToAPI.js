import { BACKEND_IP } from '../config';

function fetchToAPI(urlString, options) {
    return fetch( BACKEND_IP + urlString, options);
}

export default fetchToAPI;