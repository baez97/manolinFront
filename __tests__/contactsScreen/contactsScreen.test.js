import React from 'react';
import { Linking } from 'react-native';
import renderer from 'react-test-renderer';
import ContactsScreen from '../../screens/contactsScreen';
import ContactCard    from '../../components/contactsComponents/contactCard';
import fetchToAPI     from '../../components/fetchToAPI';

jest.unmock("react-native")
jest.unmock('react')
jest.unmock("react-test-renderer")
jest.unmock('expo');
jest.mock("../../components/fetchToAPI")

describe("ContactsScreen", () => {
    describe("Functions", () => {

        it("Constructor calls to GetParam", () => {
            const mockedGetParam = jest.fn();
            const mockedProps = {
                navigation: {
                    getParam: mockedGetParam
                }
            }
    
            new ContactsScreen(mockedProps);
            
            expect( mockedGetParam ).toBeCalled();
        })
    
        it("ComponentDidMount calls to loadContacts", () => {
            const mockedGetParam = jest.fn();
            const mockedProps = {
                navigation: {
                    getParam: mockedGetParam
                }
            }
    
            const CS = new ContactsScreen(mockedProps);
            CS.loadContacts = jest.fn();
            
            CS.componentDidMount();
    
            expect( CS.loadContacts ).toBeCalled();
        });
    
        it("ShowModal calls to SetState", () => {
            const mockedGetParam = jest.fn();
            const mockedProps = {
                navigation: {
                    getParam: mockedGetParam
                }
            }
    
            const CS = new ContactsScreen(mockedProps);
            CS.setState = jest.fn();
            
            CS.showModal("MOCKED_CONTACT");
    
            expect( CS.setState ).toBeCalled();
        });

        it("HideModal calls to SetState", () => {
            mockedGetParam = jest.fn();
            const mockedProps = {
                navigation: {
                    getParam: mockedGetParam
                }
            }
    
            const CS = new ContactsScreen(mockedProps);
            CS.setState = jest.fn();
            
            CS.hideModal();
    
            expect( CS.setState ).toBeCalled();
        });

        it("KeyExtractor gets the id of the passed object", () => {
            mockedGetParam = jest.fn();
            const mockedProps = {
                navigation: {
                    getParam: mockedGetParam
                }
            }
    
            const CS = new ContactsScreen(mockedProps);
            const object = { _id: "MOCKED_ID" }
            const result = CS.keyExtractor(object);
    
            expect( result ).toBe("MOCKED_ID");
        });

        describe("RenderItem", () => {
            it("Returns null if the name is the same", () => {
                mockedGetParam = jest.fn(() => "Ana");
                const mockedProps = {
                    navigation: {
                        getParam: mockedGetParam
                    }
                }

                const CS = new ContactsScreen(mockedProps);
                const result = CS.renderItem({item: {name: "Ana"}});

                expect( result ).toBe(null); 
            });

            it("Returns a ContactCard if the name is not the same", () => {
                mockedGetParam = jest.fn(() => "María");
                const mockedProps = {
                    navigation: {
                        getParam: mockedGetParam
                    }
                }

                const CS = new ContactsScreen(mockedProps);
                const result = CS.renderItem({item: {name: "Ana"}});

                expect( result.type ).toBe(ContactCard); 
            });
        });

        describe("LoadContactd (fetch)", () => {
            mockedGetParam = jest.fn(() => "María");
            
            it("Catches a fetch error", async () => {
                const mockedProps = {
                    navigation: {
                        getParam: mockedGetParam
                    }
                }
    
                const CS = new ContactsScreen(mockedProps);

                fetchToAPI.mockReturnValueOnce(
                    new Promise((res, rej) => {
                        rej("ERROR");
                    })
                );
                
                CS.setState = jest.fn();

                await CS.loadContacts();

                expect(CS.setState).toBeCalledWith({error: true});
            });

            it("Throws an error if the APIRest returns it", async () => {
                const mockedProps = {
                    navigation: {
                        getParam: mockedGetParam
                    }
                }
    
                const CS = new ContactsScreen(mockedProps);

                fetchToAPI.mockReturnValueOnce(
                    new Promise((res, rej) => {
                        res({
                            json: () => ({err: true})
                        });
                    })
                );

                CS.setState = jest.fn();

                await CS.loadContacts();

                expect(CS.setState).toBeCalledWith({error: true});
            });

            it("Calls to SetState if everything was fine", async () => {
                const mockedProps = {
                    navigation: {
                        getParam: mockedGetParam
                    }
                }
    
                const CS = new ContactsScreen(mockedProps);

                fetchToAPI.mockReturnValueOnce(
                    new Promise((res, rej) => {
                        res({
                            json: () => ({err: false, result: "-"})
                        });
                    })
                );

                CS.setState = jest.fn();

                const expectedArgs = {
                    contacts: "-",
                    loaded: true
                }
                await CS.loadContacts();

                expect(CS.setState).toBeCalledWith(expectedArgs);
            });
        });
    });

    describe("View", () => {
        it("Matches the Snapshot when there is an error", () => {
            const mockedProps = {
                navigation: {
                    getParam: mockedGetParam
                }
            }

            const CS = new ContactsScreen(mockedProps);
            CS.state = { error: true };

            expect( CS.render() ).toMatchSnapshot();
        });

        it("Matches the Snapshot when it is loading", () => {
            const mockedProps = {
                navigation: {
                    getParam: mockedGetParam
                }
            }

            const CS = new ContactsScreen(mockedProps);
            CS.state = { loading: true };

            expect( CS.render() ).toMatchSnapshot();
        });

        it("Matches the Snapshot when contacts were loaded", () => {
            const mockedContacts = [
                {
                    name: "Ana",
                    phone: "MOCKED_PHONE_1"
                },
                {
                    name: "María",
                    phone: "MOCKED_PHONE_2"
                }
            ];

            const mockedProps = {
                navigation: {
                    getParam: mockedGetParam
                }
            }

            const CS = new ContactsScreen(mockedProps);
            CS.state = { loaded: true, contacts: mockedContacts };

            expect( CS.render() ).toMatchSnapshot();
        });
    });
});