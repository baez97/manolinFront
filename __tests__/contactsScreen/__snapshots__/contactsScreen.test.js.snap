// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsScreen View Matches the Snapshot when contacts were loaded 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#bee6ef",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
      "paddingTop": 70,
    }
  }
>
  <Component
    allowFontScaling={false}
    style={
      Object {
        "fontFamily": "montserrat-extra-bold",
        "fontSize": 38.95833442608516,
        "textAlign": "center",
      }
    }
  >
    Contactos
  </Component>
  <FlatList
    contentContainerStyle={
      Object {
        "paddingBottom": 70,
        "paddingLeft": 35,
        "paddingRight": 35,
      }
    }
    data={
      Array [
        Object {
          "name": "Ana",
          "phone": "MOCKED_PHONE_1",
        },
        Object {
          "name": "MarÃ­a",
          "phone": "MOCKED_PHONE_2",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
  <ContactModal
    hideModal={[Function]}
  />
</Component>
`;

exports[`ContactsScreen View Matches the Snapshot when it is loading 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#bee6ef",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
      "paddingTop": 70,
    }
  }
>
  <PulseIndicator
    animationEasing={[Function]}
    color="#067bdb"
    size={262.5}
  />
</Component>
`;

exports[`ContactsScreen View Matches the Snapshot when there is an error 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#bee6ef",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
      "paddingTop": 70,
    }
  }
>
  <Component
    allowFontScaling={false}
  >
    Se ha producido un error
  </Component>
</Component>
`;
